File: workspace/2023-02-08/run-length-encoding.ruby/run_length_encoding.rb

def encode(xs)
    xs_to_encoded_runs(xs)
end

def xs_to_encoded_runs(xs)
    runs = xs_to_runs(xs)
    runs_to_encoded_runs(runs)
end

def runs_to_encoded_runs(runs)
    runs.map do |run|
        run_to_encoded_run(run)
    end
end

def run_to_encoded_run(run)
    [run[0], run.size]
end

def xs_to_runs(xs)
    if xs.empty?
        return []
    end

    runs = []
    run = xs[0..0]
    xs = xs[1..-1]

    xs.each do |x|
        if x == run[0]
            run.append(x)
        else
            runs.append(run)
            run = [x]
        end
    end

    runs.append(run)
end

def decode(encoded_runs)
    encoded_runs_to_xs(encoded_runs)
end

def encoded_runs_to_xs(encoded_runs)
    runs = encoded_runs_to_runs(encoded_runs)
    runs_to_xs(runs)
end

def encoded_runs_to_runs(encoded_runs)
    encoded_runs.map do |encoded_run|
        encoded_run_to_run(encoded_run)
    end
end

def encoded_run_to_run(encoded_run)
    encoded_run[1].times.map do
        encoded_run[0]
    end
end

def runs_to_xs(runs)
    runs.reduce([], :+)
end

File: workspace/2023/2023-02-01/run-length-encoding.python/run_length_encoding.py

def encode(xs):
    return xs_to_encoded_runs(xs)


def xs_to_encoded_runs(xs):
    runs = xs_to_runs(xs)
    return runs_to_encoded_runs(runs)


def xs_to_runs(xs):
    if len(xs) == 0:
        return []

    runs, run, xs = [], xs[:1], xs[1:]

    while len(xs) != 0:
        if xs[0] == run[0]:
            run.append(xs[0])
            xs = xs[1:]
        else:
            runs.append(run)
            run, xs = xs[:1], xs[1:]

    runs.append(run)
    return runs


def runs_to_encoded_runs(runs):
    return list(map(run_to_encoded_run, runs))


def run_to_encoded_run(run):
    return (run[0], len(run))


def decode(encoded_runs):
    return encoded_runs_to_xs(encoded_runs)


def encoded_runs_to_xs(encoded_runs):
    runs = encoded_runs_to_runs(encoded_runs)
    return runs_to_xs(runs)


def encoded_runs_to_runs(encoded_runs):
    return list(map(encoded_run_to_run, encoded_runs))


def encoded_run_to_run(encoded_run):
    run = []

    for i in range(encoded_run[1]):
        run += encoded_run[0]
    return run


def runs_to_xs(runs):
    xs = []

    for run in runs:
        xs += run
    return xs


